src/hotspot/share/runtime/thread.hpp

```diff
@@ -1687,6 +1689,7 @@ class JavaThread: public Thread {
   void verify();
   const char* get_thread_name() const;
  protected:
+  friend class JvmtiTrace;  // so get_thread_name_string() can be called
```

Comment:
I was trying to think of a way to not have JvmtiTrace not be a friend of JavaThread for this, maybe by adding a default value parameter to return "\<NOT FILLED IN\>" rather than Thread::name.
is_JavaThread_protected only seems to be called by JvmtiTrace also, so should be private (with the friend, which also makes the friend unfortunate).

Reply:
JavaThread::get_thread_name() also calls is_JavaThread_protected().

The "friend" is so that JvmtiTrace can call get_thread_name_string()
and we can get rid of the JvmtiTrace version of the logic. I kept the
"<NOT FILLED IN>" rather than figure out a way to call Thread::name()
so we don't introduce the possibility of a compatibility issue for any
code that might depend on that hand rolled string value...

Reply:
So Thread::is_JavaThread_protected() should be "protected" then, not public.
yes, I was suggesting adding a default last parameter like
    JavaThread::get_thread_name(char* default = Thread::name());
and pass "\<NOT_FILLED_IN\>" from JVMTI. Then JVMTI doesn't have to be a friend and have more visibility to the JavaThread class than it should have.

Reply:
Ahhh.... I think I finally understand what you mean...
I'll look in the AM after I've had some coffee... :-)

Reply:
The name "default" is reserved so I went with "def_name".
You can't call Thread::name() from the declaration:
```
./open/src/hotspot/share/runtime/thread.hpp:1690:62: error: call to non-static member function without an object argument
  const char* get_thread_name(const char* def_name = Thread::name()) const;
                                                     ~~~~~~~~^~~~
```
so I went with setting `def_name = NULL` in the declaration and decoding that state in get_thread_name():
` return (def_name != NULL) ? def_name : Thread::name();`

Reply:
default_name, it's not long enough to abbreviate...  but yes. thanks!

Reply:
And it looks like making is_JavaThread_protected() a protected function doesn't work either:

Reply:
`JavaThread::get_thread_name()` is calling `is_JavaThread_protected()` on the calling Thread and NOT on the `this` JavaThread so `protected` doesn't work.

Reply:
ok.